homeassistant/components/modbus/modbus.py

```diff
@@ -242,10 +243,13 @@ async def async_setup(self):
                 )
         except ModbusException as exception_error:
             self._log_error(exception_error, error_state=False)
-            return
+            return False
 
         async with self._lock:
-            await self.hass.async_add_executor_job(self._pymodbus_connect)
+            if not await self.hass.async_add_executor_job(self._pymodbus_connect):
+                text = ModbusException("initial connect failed, no retry")
```

Comment:
I don't think we should create a ModbusException instance here. Those should only be created by that library. Otherwise the origin of the exception is misleading.

We can create an exception that inherits from HomeAssistantError.

Reply:
And please use another variable name than `text`. It's not a string.

Reply:
Variable name changed. 
This is only used to log the error message, and _log_error (which are used in many places) expect a ModBusException (due to strong typing) and I do not want to change that.

This is really an exception that should come from pymodbus, since it signals no connect, just in this version someone decided not to make an exception but a return code instead (in some older version it was an exception, which is why I did not check the return code earlier).

Reply:
We shouldn't use the library exception. If we need to raise exceptions we raise our own exceptions.

Reply:
I have to agree with Martin on this one. This is a third-party exception, which should not be thrown by us. Instead, I would have expected an exception that inherits HomeAssistantError.

As for the `_log_error` method, I think that can be adjusted quite easily. The quickest method would be changing the signature, as it actually could accept any exception, so this

```py
def _log_error(self, exception_error: ModbusException, error_state=True):
```

Cloud become:

```py
def _log_error(self, exception_error: Exception, error_state=True):
```

Which fixes the typing concern issue raised. There is nothing in that method that requires more specific typing in this case. Besides, the _log_error function isn't typed as the signature is missing typing for most arguments and missing the return type.

---

Taking a small step back, it seems like `_log_error` is mainly here to prevent log spam, it actually doesn't need an exception. (And it doesn't always get an exception, see line 321, where a result is pushed into this method, which is not an exception).

IMHO, `_log_error` should be changed to:

```py
def _log_error(self, message: str, error_state: bool = True) -> None:
    if self._in_error:
		_LOGGER.debug(message)
        return
	_LOGGER.error(message)
    self._in_error = error_state
```

Now it accepts just strings. At this point, we don't even need to raise an exception. The 5 places that use this _log_error method that do pass an exception can cast it to a string directly.

e.g., `self._log_error(str(exception_error))`

I think the latter is the best and cleanest solution in general.

Reply:
why do you all continue to talk about raising an exception, there is no “raise”. I am assigning a variable to an exception, but never throwing it.

I could agree with the argumentation if I did throw the exception to be caught somewhere else.

However as usual discussion lost, so why bother trying as @french suggested to me earlier.

Reply:
> why do you all continue to talk about raising an exception, there is no “raise”. I am assigning a variable to an exception, but never throwing it.

True, bad word choice from my end, sorry.  Nevertheless, there is no reason to use an exception as a string transport.

Reply:
And if you look it is changed.

So I assume this PR can be approved, then hope the 2 other PR #54412, #54417 can be approved as well, so I can continue. There are no open review comments in the 2, and I answered why I prefer to write “30/1000” instead of “0.003” when representing milliseconds.

Reply:
Seems push is failing.